File: Kafka340/publisher-module/src/main/java/by/rest/publisher/service/CommentService.java
package by.rest.publisher.service;

import by.rest.publisher.domain.Comment;
import by.rest.publisher.dto.comment.CommentRequestTo;
import by.rest.publisher.dto.comment.CommentResponseTo;
import by.rest.publisher.dto.kafka.CommentKafkaRequest;
import by.rest.publisher.exception.ApiException;
import by.rest.publisher.mapper.CommentMapper;
import by.rest.publisher.repository.CommentRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.UUID;

@Service
@Transactional
public class CommentService {
    
    private static final Logger logger = LoggerFactory.getLogger(CommentService.class);
    
    private final CommentRepository commentRepository;
    private final CommentMapper commentMapper;
    private final KafkaProducerService kafkaProducerService;
    
    public CommentService(CommentRepository commentRepository, 
                         CommentMapper commentMapper,
                         KafkaProducerService kafkaProducerService) {
        this.commentRepository = commentRepository;
        this.commentMapper = commentMapper;
        this.kafkaProducerService = kafkaProducerService;
    }
    
    public CommentResponseTo createComment(CommentRequestTo request) {
        validateRequest(request);
        
        // Создаем сущность Comment
        Comment comment = commentMapper.toEntity(request);
        comment.setStatus("PENDING");
        
        // Сохраняем в БД
        comment = commentRepository.save(comment);
        logger.info("Comment saved to database: id={}, storyId={}, status={}", 
                comment.getId(), comment.getStoryId(), comment.getStatus());
        
        // Отправляем на модерацию через Kafka
        CommentKafkaRequest kafkaRequest = new CommentKafkaRequest(
            comment.getId().toString(), 
            request.getContent(), 
            request.getStoryId(),
            request.getAuthor() != null ? request.getAuthor() : "anonymous",
            System.currentTimeMillis()
        );
        
        kafkaProducerService.sendCommentForModeration(kafkaRequest);
        
        logger.info("Comment sent to Kafka for moderation: id={}", comment.getId());
        
        // Возвращаем DTO
        return commentMapper.toResponse(comment);
    }
    
    @Transactional(readOnly = true)
    public CommentResponseTo getCommentById(UUID id) {
        Comment comment = commentRepository.findById(id)
                .orElseThrow(() -> new ApiException(404, "40401", "Comment not found with id: " + id));
        
        return commentMapper.toResponse(comment);
    }
    
    @Transactional(readOnly = true)
    public List<CommentResponseTo> getCommentsByStoryId(Long storyId) {
        List<Comment> comments = commentRepository.findByStoryId(storyId);
        return comments.stream()
                .map(commentMapper::toResponse)
                .toList();
    }
    
    public void updateCommentStatus(UUID commentId, String status) {
        commentRepository.findById(commentId).ifPresentOrElse(
            comment -> {
                String oldStatus = comment.getStatus();
                comment.setStatus(status);
                commentRepository.save(comment);
                logger.info("Comment status updated: id={}, oldStatus={}, newStatus={}", 
                        commentId, oldStatus, status);
            },
            () -> {
                logger.warn("Cannot update status: comment not found in database, id={}", commentId);
            }
        );
    }
    
    public void deleteComment(UUID id) {
        if (!commentRepository.existsById(id)) {
            throw new ApiException(404, "40401", "Comment not found with id: " + id);
        }
        commentRepository.deleteById(id);
        logger.info("Comment deleted from database: id={}", id);
    }
    
    @Transactional(readOnly = true)
    public List<CommentResponseTo> getAllComments() {
        return commentRepository.findAll().stream()
                .map(commentMapper::toResponse)
                .toList();
    }
    
    public List<CommentResponseTo> getCommentsByStatus(String status) {
        List<Comment> comments = commentRepository.findByStatus(status);
        return comments.stream()
                .map(commentMapper::toResponse)
                .toList();
    }
    
    private void validateRequest(CommentRequestTo request) {
        if (request.getStoryId() == null || request.getStoryId() <= 0) {
            throw new ApiException(400, "40001", "Story ID must be positive");
        }
        
        if (request.getContent() == null || request.getContent().trim().isEmpty()) {
            throw new ApiException(400, "40002", "Content cannot be empty");
        }
        
        if (request.getContent().length() < 2) {
            throw new ApiException(400, "40003", "Content must be at least 2 characters long");
        }
        
        if (request.getContent().length() > 2048) {
            throw new ApiException(400, "40004", "Content cannot exceed 2048 characters");
        }
    }
    
    public long getCommentCount() {
        return commentRepository.count();
    }
    
    public long getCommentCountByStory(Long storyId) {
        return commentRepository.countByStoryId(storyId);
    }
} 

File: 351002/Bubich/t330/discussion/app/services/comment_service.py
from typing import List, Dict, Any
from app.repository.cassandra_repository import CassandraRepository
from werkzeug.exceptions import NotFound, BadRequest


class CommentService:
    def __init__(self, repository: CassandraRepository):
        self.repository = repository

    def create(self, data: Dict[str, Any]):
        self._validate(data)

        # Генерируем ID если не передан
        if 'id' not in data:
            data['id'] = self._generate_id()

        return self.repository.save(data)

    def get_by_id(self, id: int):
        comment = self.repository.find_by_id(id)
        if not comment:
            raise NotFound(f"Comment with id {id} not found")
        return comment

    def get_by_story_id(self, story_id: int) -> List[Dict]:
        return self.repository.find_by_story_id(story_id)

    def get_all(self):
        return self.repository.find_all()

    def update(self, id: int, data: Dict[str, Any]):
        self._validate(data)

        existing = self.repository.find_by_id(id)
        if not existing:
            raise NotFound(f"Comment with id {id} not found")

        return self.repository.update(id, data)

    def delete(self, id: int):
        comment = self.repository.find_by_id(id)
        if not comment:
            raise NotFound(f"Comment with id {id} not found")

        self.repository.delete_by_id(id, comment.get('story_id'))

    def _validate(self, data: Dict[str, Any]):
        content = data.get('content', '')
        if len(content) < 2 or len(content) > 2048:
            raise BadRequest("Content must be between 2 and 2048 characters")

    def _generate_id(self):
        import time
        return int(time.time() * 1000) % (10 ** 10)